File: src/main/java/org/modelingvalue/nelumbo/patterns/Pattern.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2026 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo.patterns;

import java.io.Serial;
import java.util.function.Function;

import org.modelingvalue.collections.List;
import org.modelingvalue.collections.Map;
import org.modelingvalue.collections.mutable.MutableList;
import org.modelingvalue.nelumbo.AstElement;
import org.modelingvalue.nelumbo.Node;
import org.modelingvalue.nelumbo.Type;
import org.modelingvalue.nelumbo.Variable;
import org.modelingvalue.nelumbo.syntax.ParseState;
import org.modelingvalue.nelumbo.syntax.Token;
import org.modelingvalue.nelumbo.syntax.TokenType;

public abstract class Pattern extends Node {
    @Serial
    private static final long serialVersionUID = -1788203180486332564L;

    public static Pattern a(Pattern... options) {
        return a(List.of(), options);
    }

    public static Pattern n(Type nodeType, int precedence) {
        return n(List.of(), nodeType, precedence, null);
    }

    public static Pattern n(Type nodeType, int precedence, Boolean visible) {
        return n(List.of(), nodeType, precedence, visible);
    }

    public static Pattern o(Pattern optional) {
        return o(List.of(), optional);
    }

    public static Pattern r(Pattern repeated, Boolean mandatory, Pattern separator) {
        return r(List.of(), repeated, mandatory, separator);
    }

    public static Pattern s(Pattern... elements) {
        return s(List.of(), elements);
    }

    public static Pattern t(String tokenText) {
        return t(List.of(), tokenText);
    }

    public static Pattern k(String tokenText) {
        return k(List.of(), tokenText);
    }

    public static Pattern t(Variable var) {
        return t(List.of(), var);
    }

    public static Pattern t(TokenType tokenType) {
        return t(List.of(), tokenType);
    }

    public static Pattern v(Variable var) {
        return v(List.of(), var);
    }

    public static Pattern a(List<AstElement> ast, Pattern... options) {
        return new AlternationPattern(Type.PATTERN, ast, List.of(options));
    }

    public static Pattern n(List<AstElement> ast, Type nodeType, Integer precedence, Boolean visible) {
        return new NodeTypePattern(Type.PATTERN, ast, nodeType, precedence, visible);
    }

    public static Pattern o(List<AstElement> ast, Pattern optional) {
        return new OptionalPattern(Type.PATTERN, ast, optional);
    }

    public static Pattern r(List<AstElement> ast, Pattern repeated, Boolean mandatory, Pattern separator) {
        return new RepetitionPattern(Type.PATTERN, ast, repeated, mandatory, separator);
    }

    public static Pattern s(List<AstElement> ast, Pattern... elements) {
        return new SequencePattern(Type.PATTERN, ast, List.of(elements).replaceAllAll(e -> e instanceof SequencePattern s ? s.elements() : List.of(e)));
    }

    public static Pattern t(List<AstElement> ast, String tokenText) {
        return new TokenTextPattern(Type.PATTERN, ast, tokenText, false);
    }

    public static Pattern k(List<AstElement> ast, String tokenText) {
        return new TokenTextPattern(Type.PATTERN, ast, tokenText, true);
    }

    public static Pattern t(List<AstElement> ast, Variable var) {
        return new TokenTextPattern(Type.PATTERN, ast, var, false);
    }

    public static Pattern t(List<AstElement> ast, TokenType tokenType) {
        return new TokenTypePattern(Type.PATTERN, ast, tokenType);
    }

    public static Pattern v(List<AstElement> ast, Variable var) {
        Type type = var.type();
        TokenType tt = type.tokenType();
        return tt != null ? new TokenTextPattern(Type.PATTERN, ast, var, false) : //
                new NodeTypePattern(Type.PATTERN, ast, new Type(var), null, null);
    }

    protected Pattern(Type type, List<AstElement> ast, Object... args) {
        super(type, ast, args);
    }

    protected Pattern(Object[] args, Pattern declaration) {
        super(args, declaration);
    }

    @Override
    protected abstract Pattern struct(Object[] array, Node declaration);

    public abstract ParseState state(ParseState next);

    public String name() {
        return "";
    }

    public Pattern setPresedence(int precedence) {
        return this;
    }

    public Pattern setTypes(Function<Type, Type> typeFunction) {
        return this;
    }

    @Override
    public Pattern set(int i, Object... a) {
        return (Pattern) super.set(i, a);
    }

    public abstract List<Type> argTypes(List<Type> types);

    protected abstract int string(List<Object> args, int ai, StringBuffer sb, TokenType[] previous, boolean alt);

    protected abstract int args(List<AstElement> elements, int i, MutableList<Object> args, boolean alt, Functor functor, Map<Variable, Type> typeArgs);

    public static boolean isEndOfLine(Token token) {
        return token.type() == TokenType.ENDOFFILE || (token.previous() != null && token.line() > token.previous().line());
    }

    protected void addText(StringBuffer sb, TokenType[] previous, String text) {
        TokenType type = TokenType.of(text);
        if (previous[0] == TokenType.NAME || previous[0] == TokenType.NUMBER || previous[0] == TokenType.DECIMAL) {
            if (type == TokenType.NAME || type == TokenType.NUMBER || type == TokenType.DECIMAL) {
                sb.append(" ");
            }
        }
        sb.append(text);
        previous[0] = type;
    }

    public abstract Pattern declaration(Token token);

}

File: web-ts/src/patterns/Pattern.ts
/**
 * Pattern abstract base class for syntax pattern matching.
 * @JAVA_REF org.modelingvalue.nelumbo.patterns.Pattern
 */

import { List, Map } from 'immutable';
import { TokenType } from '../syntax/TokenType';
import type { Token } from '../syntax/Token';
import type { AstElement } from '../AstElement';
import { Type } from '../Type';
import { Variable } from '../Variable';
import { Node } from '../Node';
import type { ParseState } from '../syntax/ParseState';
import type { Functor } from './Functor';

/**
 * Abstract base class for syntax patterns.
 */
export abstract class Pattern extends Node {
  protected constructor(type: Type, elements: List<AstElement>, ...args: unknown[]) {
    super(type, elements, ...args);
  }

  /**
   * Build a parse state for this pattern.
   */
  abstract parseState(next: ParseState, functor: Functor): ParseState;

  /**
   * Get the name of this pattern (for functor naming).
   */
  name(): string {
    return '';
  }

  /**
   * Set precedence for inner patterns.
   */
  setPrecedence(_precedence: number): Pattern {
    return this;
  }

  /**
   * Apply a type transformation function.
   */
  setTypes(_typeFunction: (type: Type) => Type): Pattern {
    return this;
  }

  /**
   * Get the argument types this pattern produces.
   */
  abstract argTypes(types: List<Type>): List<Type>;

  /**
   * Convert arguments to string representation.
   */
  abstract string(args: List<unknown>, ai: number, sb: string[], previous: TokenType[], alt: boolean): number;

  /**
   * Extract arguments from AST elements.
   */
  abstract extractArgs(
    elements: List<AstElement>,
    i: number,
    args: unknown[],
    alt: boolean,
    functor: Functor,
    typeArgs: Map<Variable, Type>
  ): number;

  /**
   * Get the declaration pattern for a token.
   */
  abstract tokenDeclaration(token: Token): Pattern | null;

  /**
   * Check if a token is at end of line.
   */
  static isEndOfLine(token: Token): boolean {
    return token.type === TokenType.ENDOFFILE ||
           (token.previous !== null && token.line > token.previous.line);
  }

  /**
   * Add text with proper spacing based on token types.
   */
  protected addText(sb: string[], previous: TokenType[], text: string): void {
    const type = TokenType.of(text);
    if (previous[0] === TokenType.NAME ||
        previous[0] === TokenType.NUMBER ||
        previous[0] === TokenType.DECIMAL) {
      if (type === TokenType.NAME ||
          type === TokenType.NUMBER ||
          type === TokenType.DECIMAL) {
        sb.push(' ');
      }
    }
    sb.push(text);
    previous[0] = type;
  }

  // Static factory methods

  /**
   * Create an alternation pattern.
   */
  static a(...options: Pattern[]): Pattern {
    return Pattern.aWithElements(List(), ...options);
  }

  /**
   * Create a node type pattern.
   */
  static n(nodeType: Type, precedence: number | null): Pattern {
    return Pattern.nWithElements(List(), nodeType, precedence);
  }

  /**
   * Create an optional pattern.
   */
  static o(optional: Pattern): Pattern {
    return Pattern.oWithElements(List(), optional);
  }

  /**
   * Create a repetition pattern.
   */
  static r(repeated: Pattern, mandatory: boolean, separator: Pattern | null): Pattern {
    return Pattern.rWithElements(List(), repeated, mandatory, separator);
  }

  /**
   * Create a sequence pattern.
   */
  static s(...elements: Pattern[]): Pattern {
    return Pattern.sWithElements(List(), ...elements);
  }

  /**
   * Create a token text pattern.
   */
  static t(tokenText: string): Pattern;
  static t(tokenType: TokenType): Pattern;
  static t(variable: Variable): Pattern;
  static t(value: string | TokenType | Variable): Pattern {
    if (typeof value === 'string') {
      return Pattern.tTextWithElements(List(), value, false);
    } else if (value instanceof Variable) {
      return Pattern.tVarWithElements(List(), value);
    } else {
      return Pattern.tTypeWithElements(List(), value);
    }
  }

  /**
   * Create a keyword token text pattern.
   */
  static k(tokenText: string): Pattern {
    return Pattern.kWithElements(List(), tokenText);
  }

  /**
   * Create a variable pattern.
   */
  static v(variable: Variable): Pattern {
    return Pattern.vWithElements(List(), variable);
  }

  // Factory methods with AST elements
  // These are implemented using a registry pattern to avoid circular imports

  private static _registry: {
    AlternationPattern?: new (type: Type, ast: List<AstElement>, options: List<Pattern>) => Pattern;
    NodeTypePattern?: new (type: Type, ast: List<AstElement>, nodeType: Type, precedence: number | null) => Pattern;
    OptionalPattern?: new (type: Type, ast: List<AstElement>, optional: Pattern) => Pattern;
    RepetitionPattern?: new (type: Type, ast: List<AstElement>, repeated: Pattern, mandatory: boolean, separator: Pattern | null) => Pattern;
    SequencePattern?: new (type: Type, ast: List<AstElement>, elements: List<Pattern>) => Pattern;
    TokenTextPattern?: new (type: Type, ast: List<AstElement>, text: string | Variable, isKeyword?: boolean) => Pattern;
    TokenTypePattern?: new (type: Type, ast: List<AstElement>, tokenType: TokenType) => Pattern;
  } = {};

  static registerPatternClass(name: string, clazz: unknown): void {
    (Pattern._registry as Record<string, unknown>)[name] = clazz;
  }

  static aWithElements(ast: List<AstElement>, ...options: Pattern[]): Pattern {
    const Clazz = Pattern._registry.AlternationPattern;
    if (!Clazz) throw new Error('AlternationPattern not registered');
    return new Clazz(Type.PATTERN, ast, List(options));
  }

  static nWithElements(ast: List<AstElement>, nodeType: Type, precedence: number | null): Pattern {
    const Clazz = Pattern._registry.NodeTypePattern;
    if (!Clazz) throw new Error('NodeTypePattern not registered');
    return new Clazz(Type.PATTERN, ast, nodeType, precedence);
  }

  static oWithElements(ast: List<AstElement>, optional: Pattern): Pattern {
    const Clazz = Pattern._registry.OptionalPattern;
    if (!Clazz) throw new Error('OptionalPattern not registered');
    return new Clazz(Type.PATTERN, ast, optional);
  }

  static rWithElements(ast: List<AstElement>, repeated: Pattern, mandatory: boolean, separator: Pattern | null): Pattern {
    const Clazz = Pattern._registry.RepetitionPattern;
    if (!Clazz) throw new Error('RepetitionPattern not registered');
    return new Clazz(Type.PATTERN, ast, repeated, mandatory, separator);
  }

  static sWithElements(ast: List<AstElement>, ...elements: Pattern[]): Pattern {
    const Clazz = Pattern._registry.SequencePattern;
    if (!Clazz) throw new Error('SequencePattern not registered');
    // Flatten nested sequence patterns
    const flattened = List(elements).flatMap(e => {
      if ('patternElements' in e && typeof (e as { patternElements: unknown }).patternElements === 'function') {
        return (e as unknown as { patternElements(): List<Pattern> }).patternElements();
      }
      return List([e]);
    });
    return new Clazz(Type.PATTERN, ast, flattened);
  }

  static tTextWithElements(ast: List<AstElement>, tokenText: string, isKeyword: boolean): Pattern {
    const Clazz = Pattern._registry.TokenTextPattern;
    if (!Clazz) throw new Error('TokenTextPattern not registered');
    return new Clazz(Type.PATTERN, ast, tokenText, isKeyword);
  }

  static kWithElements(ast: List<AstElement>, tokenText: string): Pattern {
    const Clazz = Pattern._registry.TokenTextPattern;
    if (!Clazz) throw new Error('TokenTextPattern not registered');
    return new Clazz(Type.PATTERN, ast, tokenText, true);
  }

  static tVarWithElements(ast: List<AstElement>, variable: Variable): Pattern {
    const Clazz = Pattern._registry.TokenTextPattern;
    if (!Clazz) throw new Error('TokenTextPattern not registered');
    return new Clazz(Type.PATTERN, ast, variable);
  }

  static tTypeWithElements(ast: List<AstElement>, tokenType: TokenType): Pattern {
    const Clazz = Pattern._registry.TokenTypePattern;
    if (!Clazz) throw new Error('TokenTypePattern not registered');
    return new Clazz(Type.PATTERN, ast, tokenType);
  }

  static vWithElements(ast: List<AstElement>, variable: Variable): Pattern {
    const type = variable.type();
    const tt = type.tokenType();
    if (tt !== null) {
      const Clazz = Pattern._registry.TokenTextPattern;
      if (!Clazz) throw new Error('TokenTextPattern not registered');
      return new Clazz(Type.PATTERN, ast, variable);
    }
    const Clazz = Pattern._registry.NodeTypePattern;
    if (!Clazz) throw new Error('NodeTypePattern not registered');
    return new Clazz(Type.PATTERN, ast, Type.fromVariable(variable), null);
  }
}
